lib/Auth/Process/UpdateUserExtSource.php

```diff
@@ -36,25 +41,75 @@ public function __construct($config, $reserved)
             );
         }
 
+        if (!isset($config['pathToKey'])) {
+            throw new Exception(
+                'perun:UpdateUserExtSource: missing mandatory configuration option \'pathToKey\'.'
+            );
+        }
+
         if (isset($config['arrayToStringConversion'])) {
             $this->attrsToConversion = (array)$config['arrayToStringConversion'];
         } else {
             $this->attrsToConversion = [];
         }
 
         $this->attrMap = (array)$config['attrMap'];
+        $this->pathToKey = $config['pathToKey'];
     }
 
     public function process(&$request)
     {
+        $id = uniqid("", true);
+
+        $dataChallenge = [
+            'id' => $id,
+            'scriptName' => self::SCRIPT_NAME
+        ];
+
+        $json = json_encode($dataChallenge);
+
+        $curlChallenge = curl_init();
+        curl_setopt($curlChallenge, CURLOPT_POSTFIELDS, $json);
+        curl_setopt($curlChallenge, CURLOPT_URL, Module::getModuleURL('perun/getChallenge.php'));
+        curl_setopt($curlChallenge, CURLOPT_RETURNTRANSFER, true);
+
+        $challenge = curl_exec($curlChallenge);
+
+        if (empty($challenge)) {
+            Logger::error('Retrieving the challenge was not successful.');
+            return;
+        }
+
+        $jwk = JWKFactory::createFromKeyFile($this->pathToKey);
+        $algorithmManager = new AlgorithmManager([new RS512()]);
+        $jwsBuilder = new JWSBuilder($algorithmManager);
+
         $data = [
             'attributes' => $request['Attributes'],
             'attrMap' => $this->attrMap,
             'attrsToConversion' => $this->attrsToConversion,
             'perunUserId' => $request['perun']['user']->getId()
         ];
 
-        $cmd = 'curl -X POST -H "Content-Type: application/json" -d \'' . json_encode($data) . '\' ' .
+        $payload = json_encode([
+            'iat' => time(),
+            'nbf' => time(),
+            'exp' => time() + 300,
+            'challenge' => $challenge,
+            'id' => $id,
+            'data' => $data
+        ]);
+
+        $jws = $jwsBuilder
+            ->create()
+            ->withPayload($payload)
+            ->addSignature($jwk, ['alg' => 'RS512'])
+            ->build();
+
+        $serializer = new CompactSerializer();
+        $token = $serializer->serialize($jws, 0);
+
+        $cmd = 'curl -X POST -H "Content-Type: application/json" -d \'' . json_encode($token) . '\' ' .
                Module::getModuleURL('perun/updateUes.php') . ' > /dev/null &';
```

Comment:
consider adding [escapeshellarg](https://www.php.net/manual/en/function.escapeshellarg.php) around `json_encode($token)` and `Module::getModuleURL('perun/updateUes.php')`

Reply:
`json_encode` may produce single quotes